### PR TITLE
refactor(devtools): use signal apis in injection and router tree

### PR DESCRIPTION
**Pull request by sheikalthaf**
Refactor the injection and router tree components to use signal apis, in future we can make the components onPush and zoneless

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by sheikalthaf** [comment]
@dgp1130 @AleksanderBodurri Can you review this PR

File: devtools/projects/ng-devtools/src/lib/devtools-tabs/injector-tree/injector-tree.component.ts [hunk 1]
```diff
@@ -209,31 +224,31 @@ export class InjectorTreeComponent {
...
   checkIfSelectedNodeStillExists(): void {
-    if (this.selectedNode === null) {
+    if (this.selectedNode() === null) {
       this.snapToRoot(this.injectorTreeGraph);
       this.snapToRoot(this.elementInjectorTreeGraph);
       return;
     }
 
-    const injector = this.selectedNode.data.injector;
+    const injector = this.selectedNode()!.data.injector;
```
Comment (dgp1130) [inline devtools/projects/ng-devtools/src/lib/devtools-tabs/injector-tree/injector-tree.component.ts:233]: Suggestion: Save `this.selectedNode()` to a temporary variable to avoid the non-null assertion.

```typescript
const selectedNode = this.selectedNode();
if (selectedNode !== null) {
  // ...
  return;
}

const injector = selectedNode.data.injector;
// ...
```

File: devtools/projects/ng-devtools/src/lib/devtools-tabs/injector-tree/injector-tree.component.ts
Comment (dgp1130) [inline devtools/projects/ng-devtools/src/lib/devtools-tabs/injector-tree/injector-tree.component.ts:1]: I'd be curious to see if there's potential improvements from migrating the core logic here to drive from signals as that would have as bigger opportunity for simplification. However that seems complicated enough to probably be worth its own PR.

File: devtools/projects/ng-devtools/src/lib/devtools-tabs/injector-tree/injector-tree.component.ts [hunk 2]
```diff
@@ -57,36 +65,41 @@ import {
...
+      write: () => {
+        this.setUpEnvironmentInjectorVisualizer();
+        this.setUpElementInjectorVisualizer();
+        this.init();
+      },
+    });
   }
 
-  ngOnInit() {
+  private init() {
```
Comment (dgp1130) [inline devtools/projects/ng-devtools/src/lib/devtools-tabs/injector-tree/injector-tree.component.ts:99]: Question: What is the need for a separate `init()` function over `ngOnInit`?

It doesn't seem like the `this._messageBus.on` calls would cause a problem. Previously the `this._messageBus.on` calls happened prior to setting up the visualizers in `ngAfterViewInit`. I don't see a reason to go out of our way to change that ordering?

File: devtools/projects/ng-devtools/src/lib/devtools-tabs/router-tree/router-tree.component.ts [hunk 3]
```diff
@@ -16,47 +24,50 @@ import {Events, MessageBus, Route} from 'protocol';
...
+  routes = input<Route[]>([]);
 
-  private _routes: Route[] = [];
   private tree!: d3.TreeLayout<{}>;
   private tooltip: any;
+  private _messageBus = inject<MessageBus<Events>>(MessageBus);
 
-  constructor(private _messageBus: MessageBus<Events>) {}
+  constructor() {
+    effect(() => this.render());
```
Comment (dgp1130) [inline devtools/projects/ng-devtools/src/lib/devtools-tabs/router-tree/router-tree.component.ts:38]: Nit: Don't propagate the return value of `this.render()` as it may be unexpectedly consumed by `effect` if these functions change in the future.

```typescript
effect(() => {
  this.render();
});
```

File: devtools/projects/ng-devtools/src/lib/devtools-tabs/injector-tree/injector-tree.component.ts (hunk 1 above)
Comment (sheikalthaf) [reply devtools/projects/ng-devtools/src/lib/devtools-tabs/injector-tree/injector-tree.component.ts]: Done

File: devtools/projects/ng-devtools/src/lib/devtools-tabs/injector-tree/injector-tree.component.ts (hunk 2 above)
Comment (sheikalthaf) [reply devtools/projects/ng-devtools/src/lib/devtools-tabs/injector-tree/injector-tree.component.ts:99]: Sorry, it is a mistake from myside. Yes init should be called before the visualizer.

I'm trying to use latest Angular API for devtools.

File: devtools/projects/ng-devtools/src/lib/devtools-tabs/router-tree/router-tree.component.ts (hunk 3 above)
Comment (sheikalthaf) [reply devtools/projects/ng-devtools/src/lib/devtools-tabs/router-tree/router-tree.component.ts]: Done

File: devtools/projects/ng-devtools/src/lib/devtools-tabs/injector-tree/injector-tree.component.ts
Comment (sheikalthaf) [reply devtools/projects/ng-devtools/src/lib/devtools-tabs/injector-tree/injector-tree.component.ts:1]: Yes, signal brings lots of simplification to the reactive data handling. 

As a first step converting all the components to use signal API
Then second step is to mark all the components 'OnPush'
Then third step make it zoneless.

Comment (dgp1130) [reply devtools/projects/ng-devtools/src/lib/devtools-tabs/injector-tree/injector-tree.component.ts:1]: Yep sounds good, definitely worth splitting up this work.

**Review by dgp1130 — APPROVED** [review]
@AleksanderBodurri, any concerns / suggestions here?

File: devtools/projects/ng-devtools/src/lib/devtools-tabs/injector-tree/injector-providers.component.ts [hunk 4]
```diff
@@ -42,8 +42,8 @@ import {Events, MessageBus, SerializedInjector, SerializedProviderRecord} from '
           }
         </mat-select>
       </mat-form-field>
-      @if (providers.length > 0) {
-      <table mat-table [dataSource]="providers" class="mat-elevation-z4">
+      @if (providers().length > 0) {
```
Comment (AleksanderBodurri) [inline devtools/projects/ng-devtools/src/lib/devtools-tabs/injector-tree/injector-providers.component.ts:45]: While reviewing this PR I noticed this line and the line below it have a bug (not introduced in this PR). Can we update this to:

```suggestion
      @if (visibleProviders().length > 0) {
```

File: devtools/projects/ng-devtools/src/lib/devtools-tabs/injector-tree/injector-providers.component.ts [hunk 5]
```diff
@@ -42,8 +42,8 @@ import {Events, MessageBus, SerializedInjector, SerializedProviderRecord} from '
...
+      <table mat-table [dataSource]="providers()" class="mat-elevation-z4">
```
Comment (AleksanderBodurri) [inline devtools/projects/ng-devtools/src/lib/devtools-tabs/injector-tree/injector-providers.component.ts:46]:
```suggestion
      <table mat-table [dataSource]="visibleProviders()" class="mat-elevation-z4">
```

**Review by AleksanderBodurri — CHANGES_REQUESTED** [review]
Looks great, left some comments with two quick changes to fix a bug 🙏

File: devtools/projects/ng-devtools/src/lib/devtools-tabs/injector-tree/injector-providers.component.ts (hunk 4 above)
Comment (sheikalthaf) [reply devtools/projects/ng-devtools/src/lib/devtools-tabs/injector-tree/injector-providers.component.ts]: Done

File: devtools/projects/ng-devtools/src/lib/devtools-tabs/injector-tree/injector-providers.component.ts (hunk 5 above)
Comment (sheikalthaf) [reply devtools/projects/ng-devtools/src/lib/devtools-tabs/injector-tree/injector-providers.component.ts]: Done

**Comment by sheikalthaf** [comment]
@AleksanderBodurri Updated the PR with the changes suggested.

**Review by AleksanderBodurri — APPROVED** [review]
Thank you @sheikalthaf 🙏

**Comment by thePunderWoman** [comment]
This PR was merged into the repository by commit 49884518a0391668f4e20a27b371513431917562.

The changes were merged into the following branches: main, 18.1.x
